lighthouse-core/report/html/renderer/code-snippet-renderer.js

```diff
@@ -0,0 +1,260 @@
+   * @param {DetailsRenderer} detailsRenderer
+   * @param {function} toggleExpandedFn
+   * @return {DocumentFragment}
+   */
+  static renderHeader(dom, tmpl, details, detailsRenderer, toggleExpandedFn) {
+    const {lineCount, title} = details;
+    const showAll = lineCount <= 4;
+
+    const header = dom.cloneTemplate('#tmpl-lh-code-snippet__header', tmpl);
+    dom.find('.lh-code-snippet__title', header).textContent = title;
+
+    const {codeSnippetCollapse, codeSnippetExpand} = Util.UIStrings;
+    dom.find('.' + SHOW_IF_EXPANDED_CLASS, header).textContent = codeSnippetCollapse;
+    dom.find('.' + SHOW_IF_COLLAPSED_CLASS, header).textContent = codeSnippetExpand;
+
+    const toggleShowAllButton = dom.find('.lh-code-snippet__toggle-show-all', header);
+    if (showAll) {
+      toggleShowAllButton.remove();
+    } else {
+      toggleShowAllButton.addEventListener('click', () => toggleExpandedFn());
+    }
+
+    const nodeContainer = dom.find('.lh-code-snippet__node', header);
+    if (details.node) {
+      // todo: only do this if isdevtools (and check that it works fine there)
+      nodeContainer.appendChild(detailsRenderer.renderNode(details.node));
+    }
+
+    return header;
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {{content: string, number: number | string,truncated?: boolean}} line
+   * @param {LineOptions} lineOptions
+   * @return {Element}
+   */
+  static renderLine(dom, tmpl, line, lineOptions = {}) {
+    const {content, number, truncated} = line;
+
+    const clonedTemplate = dom.cloneTemplate('#tmpl-lh-code-snippet__line', tmpl);
+    const codeLine = dom.find('.lh-code-snippet__line', clonedTemplate);
+
+    if (lineOptions.highlight) {
+      codeLine.classList.add('lh-code-snippet__line--highlighted');
+    }
+    if (lineOptions.highlightMessage) {
+      codeLine.classList.add('lh-code-snippet__line--highlight-message');
+    }
+    if (lineOptions.code) {
+      codeLine.classList.add('lh-code-snippet__line--code');
+    }
+    if (lineOptions.collapsedOnly) {
+      codeLine.classList.add(SHOW_IF_COLLAPSED_CLASS);
+    }
+    if (lineOptions.expandedOnly) {
+      codeLine.classList.add(SHOW_IF_EXPANDED_CLASS);
+    }
+
+    const lineContent = content + (truncated ? '…' : '');
+    const lineContentEl = dom.find('.lh-code-snippet__line code', codeLine);
+    if (lineOptions.convertMarkdownLinkSnippets) {
+      lineContentEl.appendChild(dom.convertMarkdownLinkSnippets(lineContent));
+    } else {
+      lineContentEl.textContent = lineContent;
+    }
+
+    dom.find('.lh-code-snippet__line-number', codeLine).textContent = number.toString();
+
+    return codeLine;
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippetHighlight} highlight
+   * @return {Element}
+   */
+  static renderHighlightMessage(dom, tmpl, highlight) {
+    return CodeSnippetRenderer.renderLine(dom, tmpl, {
```

Comment:
this is when it clicked for me that `renderLine` doesn't mean `render a line of code` but just `render a line of content in our DOM`. not sure how to make that clearer earlier, maybe just jsdoc? or if every call had the `contentType: 'code'|'highlightMessage'|'empty'` enum suggestion I mentioned :D